test/test_invalid_headers.py

```diff
@@ -296,7 +296,9 @@ def test_headers_event_skipping_validation(self, frame_factory, headers):
         c.send_headers(1, headers)
 
         # Ensure headers are still normalized.
-        norm_headers = h2.utilities.normalize_outbound_headers(headers, None, False)
+        norm_headers = h2.utilities.normalize_outbound_headers(
+            h2.utilities.utf8_encode_headers(headers), None, False
+        )
```

Comment:
I'm contemplating making the encoding of headers part of the normalization pipeline, instead of being the responsibility of the outside caller.

What do you think about integrating it here?

Reply:
That's exactly what I wanted but unfortunately we have an option to bypass normalization, which breaks that neat abstraction :(

Reply:
`utf8_encode_headers` is currently called in two places: `connections.py - push_stream` and `stream.py - send_headers`. Both share a code path with `H2Stream._build_headers_frames`.

I'll try to refactor it and share the update to discuss.

Reply:
Sounds great!